### DISCUSSION TITLE
How to handle KeyboardInterrupt

### DISCUSSION THREAD

**Discussion by JesseFL** [mkleehammer/pyodbc#1056] [Q&A]
I'll start by disclosing that I am very rusty in Python, so I may be making a simple mistake.

I am trying to catch a KeyboardInterrupt while pyodbc is executing several sql statements.  I have tried both a SIGINT handler and a try/except block and neither seem to work.  My script is doing:

```
c.execute(sql01)
# do some stuff with the results
c.execute(sql02)
# do some stuff with the results
...
c.execute(sql99)
# do some stuff with the results
```
If the user is seeing unwanted results, I want them to be able to cancel and not have to wait for all the sql statements to run.

I've tried:
```
try:
    c.execute(sql01)
except KeyboardInterrupt:
    print('Process interrupted. Will now safely exit.')
    #do some other closure stuff
    connection.close()
    exit(1)
```
I've also tried a SIGINT handler.  In both cases I'm getting errors.
```
Traceback (most recent call last):
  File "C:\Users\user\AppData\Local\Programs\Python\Python39\lib\hashlib.py", line 152, in __hash_new
    def __hash_new(name, data=b'', **kwargs):
  File "_pydevd_bundle/pydevd_cython.pyx", line 1366, in _pydevd_bundle.pydevd_cython.SafeCallWrapper.__call__
  File "_pydevd_bundle/pydevd_cython.pyx", line 1667, in _pydevd_bundle.pydevd_cython.ThreadTracer.__call__
  File "c:\Users\user\.vscode\extensions\ms-python.python-2022.4.1\pythonFiles\lib\python\debugpy\_vendored\pydevd\_pydev_bundle\pydev_is_thread_alive.py", line 9, in is_thread_alive
    def is_thread_alive(t):
KeyboardInterrupt

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
... (redacted for formatting)
SystemError: <built-in function connect> returned a result with an error set
```

How do I properly handle a keyboard interrupt with pyodbc?

**Comment by maxmetzger** [comment]
I'm having a similar issue, I am running pyodbc in a docker container via kubernetes and would like to gracefully handle the `SIGINT` / `SIGHUP` signals that kubernetes sends when it wants to terminate a pod, but I haven't been able to change how pyodbc handles the signal. More details here: https://stackoverflow.com/questions/74828731/when-kuberentes-terminates-a-pod-it-kills-any-ongoing-pyodbc-calls-even-when-t
